crates/ra_hir/src/ty/infer.rs

```diff
@@ -72,6 +93,9 @@ impl InferenceResult {
     pub fn field_resolution(&self, expr: ExprId) -> Option<StructField> {
         self.field_resolutions.get(&expr).map(|it| *it)
     }
+    pub fn assoc_resolutions(&self, id: ExprOrPatId) -> Option<ImplItem> {
```

Comment:
Alternatively, we can make `ExprOrPatId`  type private, and expose two public functions, for `PatId` and `ExprId`.